[pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25896 [pytorch] update build_android.sh to not build host protoc for libtorch
* #25894 [pytorch] remove pthreadpool dependency in aten/CMake
* #25817 [pytorch][mobile] disable backward function codegen if disable_autograd is set
* #25816 [pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute
* #25697 [pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile
* **#25815 [pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH**

Summary:
Don't need call these global registerers when USE_STATIC_DISPATCH is
set as they will keep all aten functions at link time.

Should solely rely on jit/generated/register_aten_ops* to keep "interface"
aten functions (which are directly called from JIT), and rely on
STATIC_DISPATCH + linker to keep all other aten functions that are
transitively needed by the "interface" functions.

Test Plan:
- build and run in the demo app;
- with stacked diff to shrink registered "interface" functions, linker
  can strip out unused aten implementations;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25815

Differential Revision: [D17247236](https://our.internmc.facebook.com/intern/diff/D17247236)

ljk53 committed:
[pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH

Summary:
Don't need call these global registerers when USE_STATIC_DISPATCH is
set as they will keep all aten functions at link time.

Should solely rely on jit/generated/register_aten_ops* to keep "interface"
aten functions (which are directly called from JIT), and rely on
STATIC_DISPATCH + linker to keep all other aten functions that are
transitively needed by the "interface" functions.

Test Plan:
- build and run in the demo app;
- with stacked diff to shrink registered "interface" functions, linker
  can strip out unused aten implementations;

ljk53 committed:
Update on "[pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH"


Summary:
Don't need call these global registerers when USE_STATIC_DISPATCH is
set as they will keep all aten functions at link time.

Should solely rely on jit/generated/register_aten_ops* to keep "interface"
aten functions (which are directly called from JIT), and rely on
STATIC_DISPATCH + linker to keep all other aten functions that are
transitively needed by the "interface" functions.

Test Plan:
- build and run in the demo app;
- with stacked diff to shrink registered "interface" functions, linker
  can strip out unused aten implementations;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25815

Differential Revision: [D17247236](https://our.internmc.facebook.com/intern/diff/D17247236)

smessmer reviewed (changes requested):
I don't think relying on `jit/generated/register_aten_ops*` will work. Let's talk about this offline.

smessmer reviewed (approved):
re-accepting, this diff by itself is fine, but we should talk about the registration thing ^^

ljk53 committed:
Update on "[pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH"


Summary:
Don't need call these global registerers when USE_STATIC_DISPATCH is
set as they will keep all aten functions at link time.

Should solely rely on jit/generated/register_aten_ops* to keep "interface"
aten functions (which are directly called from JIT), and rely on
STATIC_DISPATCH + linker to keep all other aten functions that are
transitively needed by the "interface" functions.

Test Plan:
- build and run in the demo app;
- with stacked diff to shrink registered "interface" functions, linker
  can strip out unused aten implementations;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25815

Differential Revision: [D17247236](https://our.internmc.facebook.com/intern/diff/D17247236)

ljk53 committed:
Update on "[pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH"


Summary:
Don't need call these global registerers when USE_STATIC_DISPATCH is
set as they will keep all aten functions at link time.

Should solely rely on jit/generated/register_aten_ops* to keep "interface"
aten functions (which are directly called from JIT), and rely on
STATIC_DISPATCH + linker to keep all other aten functions that are
transitively needed by the "interface" functions.

Test Plan:
- build and run in the demo app;
- with stacked diff to shrink registered "interface" functions, linker
  can strip out unused aten implementations;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25815

Differential Revision: [D17247236](https://our.internmc.facebook.com/intern/diff/D17247236)
